tfexec/terraform.go

```diff
@@ -96,6 +101,13 @@ type TerraformCLI interface {
 	// because the terraform state rm command doesn't have -state-out option.
 	StateRm(ctx context.Context, state *State, addresses []string, opts ...string) (*State, error)
 
+	// StateReplaceProvider replaces a provider from source to destination address.
+	// If a state argument is given, use it for the input state.
+	// It returns the given state.
+	// Unlike other state subcommands, the terraform state replace-provider
+	// command doesn't support a -state-out option; it only supports the -state option.
```

Comment:
I tried it, too, and confirmed that the -state-out flag was not defined.
It is a Terraform CLI bug because the documentation and implementation do not match.
https://developer.hashicorp.com/terraform/cli/commands/state/replace-provider
https://github.com/hashicorp/terraform/blob/v1.5.5/internal/command/state_replace_provider.go

```
a05728d678f2:/work/tmp/pr-145/dir1# terraform -v
Terraform v1.5.5
on linux_arm64
+ provider registry.terraform.io/hashicorp/null v3.2.1

a05728d678f2:/work/tmp/pr-145/dir1# terraform state replace-provider -state=tmp.tfstate -state-out=out.tfstate hashicorp/null minamijoyo/null
Error parsing command-line flags: flag provided but not defined: -state-out

Usage: terraform [global options] state replace-provider [options] FROM_PROVIDER_FQN TO_PROVIDER_FQN

  Replace provider for resources in the Terraform state.

Options:

  -auto-approve           Skip interactive approval.

  -lock=false             Don't hold a state lock during the operation. This is
                          dangerous if others might concurrently run commands
                          against the same workspace.

  -lock-timeout=0s        Duration to retry a state lock.

  -ignore-remote-version  A rare option used for the remote backend only. See
                          the remote backend documentation for more information.

  -state, state-out, and -backup are legacy options supported for the local
  backend only. For more information, see the local backend's documentation.
```

Reply:
> It is a Terraform CLI bug because the documentation and implementation do not match.

That is my sense too, though I haven't fully researched the backstory.